Lwm2m asynchronous socket io

huelsenfruchtzwerg opened the pull request:
This restructures the lwm2m_engine to use a non-blocking socket access instead of the
previously used blocking style, and eliminates any socket-access from outside of the
main work loop.

The main motivation behind this is an issue within nordics nrf_modem_lib/modem-fw on nrf9160,
that leads to socket send() calls to block indefinitely when the shared memory used for
rpc-communication with the modem is already exhausted because of incoming data.

This lead to the lwm2m_engine locking up on send calls when there is also
a large amount of incoming data.

This works around this issue, by only issuing send calls when poll reports the socket to
be ready for sending, and (more importantly) by always receiving all buffered
incoming data before sending anything.

There might still be a (perhaps academic) possibility where this situation might be triggered,
when  the scheduler interrupts the lwm2m thread in-between receiving and sending, but for now
we have not yet observed this.

Besides working around the aforementioned issue, this also simplifies the way resends are
handled as they are no longer send from the main system-workqueue, and limits all interaction
with the sockets to a single thread.

huelsenfruchtzwerg committed:
lwm2m: use asynchronous socket io

This restructures the lwm2m_engine to use a non-blocking socket access
instead of the previously used blocking style, and eliminates any
socket-access from outside of the main work loop.

The main motivation behind this is an issue within nordics
nrf_modem_lib/modem-fw on nrf9160, that leads to socket send() calls to
block indefinitely when the shared memory used for
rpc-communication with the modem is already exhausted because of
incoming data.

This lead to the lwm2m_engine locking up on send calls when there is
also a large amount of incoming data.

This works around this issue, by only issuing send calls when poll
reports the socket to be ready for sending, and (more importantly) by
always receiving all buffered incoming data before sending anything.

There might still be a (perhaps academic) possibility where this
situation might be triggered, when  the scheduler interrupts the lwm2m
thread in-between receiving and sending, but for now we have not yet
observed this.

Besides working around the aforementioned issue, this also simplifies
the way resends are handled as they are no longer send from the main
system-workqueue, and limits all interaction with the sockets to a
single thread.

Signed-off-by: Henning Fleddermann <henning.fleddermann@grandcentrix.net>

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4628,28 +4534,129 @@ void lwm2m_socket_del(struct lwm2m_ctx *ctx)
+
+	from_addr_len = sizeof(from_addr);
+	len = recvfrom(client_ctx->sock_fd, in_buf, sizeof(in_buf) - 1,
+		       MSG_DONTWAIT, &from_addr, &from_addr_len);
```
nit: The socket was already put in the `O_NONBLOCK` state, `MSG_DONTWAIT` is redundant here.

huelsenfruchtzwerg replied:
Good catch, was a leftover from earlier modifications.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4628,28 +4534,129 @@ void lwm2m_socket_del(struct lwm2m_ctx *ctx)
+	int rc;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(observer_list, obs, node) {
+		if (obs->event_timestamp > obs->last_timestamp &&
```
This does not match the original condition:
```
		if (obs->event_timestamp > obs->last_timestamp &&
		    (obs->min_period_sec == 0 ||
		     timestamp > obs->last_timestamp +
				MSEC_PER_SEC * obs->min_period_sec)) {
```
Note the `obs->min_period_sec == 0` check missing.

Also please keep the original comments, the code is easier to understand that way.

huelsenfruchtzwerg replied:
Not to be snippy, but which comments do you refer to: the ones that literally repeat the condition in the if on the next line, or the ones that describe something that doesn't happen anywhere near the comment?

rlubos replied:
Ok, I see your point, the original comments are a bit bloated indeed. I was referring to the `if` conditions, which are not very approachable IMO w/o short explainiation what exactly is being checked (especially the first one) but we can live w/o comments as well I guess.

huelsenfruchtzwerg replied:
How about extracting the checks to seperate functions with names?

rlubos replied:
Sound fine to me

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4628,28 +4534,129 @@ void lwm2m_socket_del(struct lwm2m_ctx *ctx)
+				return;
+			}
+			obs->last_timestamp = timestamp;
+		} else if (timestamp >
```
This does not match the original condition:
```
		} else if (obs->max_period_sec > 0 &&
			   timestamp > obs->last_timestamp +
				MSEC_PER_SEC * obs->max_period_sec) {
```
Note the `obs->max_period_sec > 0` check missing.

huelsenfruchtzwerg replied:
Also a carry over from the earlier version this was originally based on, good catch :+1:

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4628,28 +4534,129 @@ void lwm2m_socket_del(struct lwm2m_ctx *ctx)
+		sock_fds[i].events = POLLIN;
+		sock_fds[i].revents = 0;
+	}
+	for (int i = 0; i < CONFIG_LWM2M_ENGINE_MAX_MESSAGES; ++i) {
```
Up for discussion, but I'd rather see `k_fifo` being used here, instead of looping over all allocated LwM2M messages every time. We could keep `k_fifo` per lwm2m_context (and thus socket) and then:
1. Put a message to the FIFO in `lwm2m_send_message_async()`,
2. Verify if the FIFO IS empty here, to decide whether to add POLLOUT,
3. Retrieve the message from the FIFO in `socket_send_message()`

All that is needed to make `lwm2m_message` FIFO ready is to add `void *fifo_reserved` at the beginning of the structure.

huelsenfruchtzwerg replied:
Yeah, that's something I was hesitant about as well. I decided to go with the simple stupid approach first, and it seemed "fast enough" :tm: to me, but I see your point.

huelsenfruchtzwerg replied:
I think a fifo might even be a little heavy-handed, since that's primarly used to synchronize between threads, but all producers are running in the same thread, so a slist might be enough. I'm curious what your opinion on that would be.

huelsenfruchtzwerg replied:
I chose to go with a slist for now.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4204,102 +4193,77 @@ static void lwm2m_udp_receive(struct lwm2m_ctx *client_ctx,
-		goto next_locked;
-	}
+		expiry = p->t0 + p->timeout;
+		if (expiry < timestamp) {
```
You should rather check if `((int32_t)(expiry - timestamp) < 0)`, otherwise it won't work on uint32_t overflow (i. e. when the `expiry` is near to UINT32_MAX, but `timestamp` has already overflown.

huelsenfruchtzwerg replied:
Good point.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4204,102 +4193,77 @@ static void lwm2m_udp_receive(struct lwm2m_ctx *client_ctx,
+				coap_pending_clear(p);
+				continue;
+			}
+			if (!p->retries) {
```
I'm wondering, why not `coap_pending_cycle(p)` here? It should report back whether the message has expired or not.

huelsenfruchtzwerg replied:
The messsage is already cycled in lwm2m_send_message, so this would cycle it twice in the case where it's not a timeout.

rlubos reviewed (commented):
Some early feedback.

huelsenfruchtzwerg commented on subsys/net/lib/lwm2m/lwm2m_rd_client.c:
```diff
@@ -123,6 +123,7 @@ struct lwm2m_rd_client_info {
  * documented in the LwM2M specification.
  */
 static char query_buffer[MAX(32, sizeof("ep=") + CLIENT_EP_LEN)];
+static uint8_t client_data[CONFIG_LWM2M_RD_CLIENT_MAX_DATA_LEN]; /* allocate some data for the RD */
```
This was indeed based on an earlier version originally and then rebased. Looks like I was a little careless when resolving conflicts, sorry about that.

huelsenfruchtzwerg committed:
lwm2m: keep track of observations per client

this has a number of advantages:
- allows to only create notifications for each client if there are no
  messages already waiting to be send, in practice prioritizing the
  memory for messages for answers, thus staying more "responsive".
- saves a fair bit of memory by eliminationg now redundant client_ctx
  pointer per observer.
- fixes a potential subtle bug: previously, an observer reset would've
  stopped the first observation found with a matching token, which
  might've belonged to a differen client.

Signed-off-by: Henning Fleddermann <henning.fleddermann@grandcentrix.net>

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4468,22 +4453,15 @@ static int32_t lwm2m_engine_service(const int64_t timestamp)
 
 	/* Remove observes for this context */
-	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&engine_observer_list, obs, tmp,
+	SYS_SLIST_FOR_EACH_CONTAINER(&client_ctx->observer, obs,
```
We should keep `SYS_SLIST_FOR_EACH_CONTAINER_SAFE` as we're removing the objects from the list in the loop (SAFE in this case is not about thread-safety but about safe removal of the list nodes from within the loop). Otherwise you break the loop on the first iteration as you clear the `next` pointer.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4468,22 +4453,15 @@ static int32_t lwm2m_engine_service(const int64_t timestamp)
+	SYS_SLIST_FOR_EACH_CONTAINER(&client_ctx->observer, obs,
 					  node) {
-		if (obs->ctx == client_ctx) {
-			sys_slist_remove(&engine_observer_list, prev_node,
```
I'd keep the explicit `sys_slist_remove` instead of assuming that memset is enough.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4666,11 +4647,18 @@ static void socket_loop(void)
-		if (next_retransmit < timeout) {
-			timeout = next_retransmit;
+		for (i = 0; i < sock_nfds; ++i) {
+			if (!sock_ctx[i]->pending_sends.head) {
```
Please use `sys_slist_is_empty()` instead of relying on syslist internals.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4666,11 +4647,18 @@ static void socket_loop(void)
+					timeout = next_retransmit;
+				}
+			}
+			if (!sock_ctx[i]->pending_sends.head) {
```
Same here.

rlubos reviewed (commented):
Left a few more comments regarding the slist usage. I think it'd make sense to squash the first two commits, no reason to introduce one sending mechanism just to reowrk it in the next commit. And please use the full `net: lwm2m:` prefix in the commit titles.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -4628,28 +4548,136 @@ void lwm2m_socket_del(struct lwm2m_ctx *ctx)
 		 */
-		if (poll(sock_fds, sock_nfds, lwm2m_engine_service()) < 0) {
+		rc = poll(sock_fds, sock_nfds, timeout);
+		if (poll < 0) {
```
```suggestion
if (rc < 0) {
```

huelsenfruchtzwerg replied:
:facepalm:

rlubos reviewed (commented):
One more issue spotted

huelsenfruchtzwerg commented:
accidentally pushed the wrong branch, I think this caused this label/reviewer cascade. sorry. :facepalm:

rlubos reviewed (approved):
LGTM, thanks
